File: scala3/src/main/scala/io/github/sps23/designpatterns/factory/NotificationFactory.scala
package io.github.sps23.designpatterns.factory

sealed trait Notification:
  def send(recipient: String, message: String): String

case object EmailNotification extends Notification:
  override def send(recipient: String, message: String): String =
    s"EMAIL to $recipient: $message"

case object SmsNotification extends Notification:
  override def send(recipient: String, message: String): String =
    s"SMS to $recipient: $message"

case object PushNotification extends Notification:
  override def send(recipient: String, message: String): String =
    s"PUSH to $recipient: $message"

object NotificationFactory:
  def create(channel: String): Notification =
    Option(channel)
      .map(_.trim.toLowerCase)
      .filter(_.nonEmpty)
      .map:
        case "email" => EmailNotification
        case "sms"   => SmsNotification
        case "push"  => PushNotification
        case other =>
          throw new IllegalArgumentException(s"Unsupported notification channel: $other")
      .getOrElse(throw new IllegalArgumentException("Notification channel must not be blank"))

File: kotlin/src/main/kotlin/io/github/sps23/designpatterns/factory/NotificationFactory.kt
package io.github.sps23.designpatterns.factory

sealed interface Notification {
    fun send(
        recipient: String,
        message: String,
    ): String
}

object EmailNotification : Notification {
    override fun send(
        recipient: String,
        message: String,
    ): String = "EMAIL to $recipient: $message"
}

object SmsNotification : Notification {
    override fun send(
        recipient: String,
        message: String,
    ): String = "SMS to $recipient: $message"
}

object PushNotification : Notification {
    override fun send(
        recipient: String,
        message: String,
    ): String = "PUSH to $recipient: $message"
}

object NotificationFactory {
    fun create(channel: String?): Notification {
        val normalized =
            channel
                ?.trim()
                ?.lowercase()
                ?.takeIf { it.isNotEmpty() }
                ?: throw IllegalArgumentException("Notification channel must not be blank")

        return when (normalized) {
            "email" -> EmailNotification
            "sms" -> SmsNotification
            "push" -> PushNotification
            else -> throw IllegalArgumentException("Unsupported notification channel: $normalized")
        }
    }
}
